src/buildApplication/middlewareManifest.ts

```diff
@@ -0,0 +1,137 @@
+/**
+ * IMPORTANT: this file relates to the .next/server/middleware-manifest.json file
+ *	moving forward we want to rely solely on the .vercel/output file so everything in this file
+ *  should be refactored to use .vercel/output instead as soon as possible
+ */
+
+import { readFile } from 'fs/promises';
+
+export type MiddlewareManifest = {
+	middleware: Record<
+		string,
+		{
+			env: string[];
+			files: string[];
+			name: string;
+			matchers: { regexp: string }[];
+			wasm: [];
+			assets: [];
+		}
+	>;
+	functions: Record<
+		string,
+		{
+			env: string[];
+			files: string[];
+			name: string;
+			page: string;
+			matchers: { regexp: string }[];
+			wasm: [];
+			assets: [];
+		}
+	>;
+	version: 2;
+};
+
+export type MiddlewareManifestData = Awaited<
+	ReturnType<typeof parseMiddlewareManifest>
+>;
+
+/**
+ * Parses the middleware manifest and validates it against the existing functions map.
+ */
+export async function parseMiddlewareManifest(
+	functionsMap: Map<string, string>
+) {
+	let middlewareManifest: MiddlewareManifest;
+	try {
+		// Annoying that we don't get this from the `.vercel` directory.
+		// Maybe we eventually just construct something similar from the `.vercel/output/functions` directory with the same magic filename/precendence rules?
+		middlewareManifest = JSON.parse(
+			await readFile('.next/server/middleware-manifest.json', 'utf8')
```

Comment:
is this along the lines of what you head in mind? 🙂 

https://github.com/cloudflare/next-on-pages/pull/127/commits/a71ea2498e68af59808b6a4879f3877fa40a924d

(PS: I am trying to keep things in the middlewareManifest as isolated as I can since we should refactor its logic at some point, that's why I kept the file reading in this file)